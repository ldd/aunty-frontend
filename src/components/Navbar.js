import React, { useState, useCallback } from "react";

export const NavBarBurger = ({ menuStatus }) => (
  <span
    className={`navbar-burger burger ${menuStatus}`}
    data-target="navbarMenuHeroC"
  >
    {/* needed to simulate a burger menu */}
    <span />
    <span />
    <span />
  </span>
);

export const NavBarLink = ({ changeRoute, route, link }) => (
  <a
    href={`#${link}`}
    onClick={changeRoute}
    className={`navbar-item ${link === route ? "is-active" : ""}`}
  >
    {link}
  </a>
);

const Navbar = ({ route, changeRoute, pages = [] }) => {
  const [menuStatus, toggleMenuStatus] = useState("");

  const menuHandler = useCallback(() => {
    toggleMenuStatus(menuStatus === "" ? "is-active" : "");
  }, [toggleMenuStatus, menuStatus]);

  const generateChangeRoute = link => () => {
    toggleMenuStatus("");
    changeRoute(link);
  };
  return (
    <div className="hero-head">
      <header className="navbar">
        <div className="container">
          <div className="navbar-brand" onClick={menuHandler}>
            <NavBarBurger menuStatus={menuStatus} />
          </div>
          <div id="navbarMenuHeroC" className={`navbar-menu ${menuStatus}`}>
            <div className="navbar-end">
              {pages.map(link => (
                <NavBarLink
                  link={link}
                  route={route}
                  changeRoute={generateChangeRoute(link)}
                  key={`navbar-${link}`}
                />
              ))}
            </div>
          </div>
        </div>
      </header>
    </div>
  );
};
export default Navbar;
