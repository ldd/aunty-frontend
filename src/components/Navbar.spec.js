import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import Navbar, { NavBarLink, NavBarBurger } from "./Navbar";

describe("Navbar", () => {
  it("renders snapshot [empty]", () => {
    const component = renderer.create(<Navbar />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders snapshot [pages]", () => {
    const pages = ["pageOne", "pageTwo"];
    const component = renderer.create(<Navbar pages={pages} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("should render no empty navbar", () => {
    const wrapper = shallow(<Navbar />);
    expect(wrapper.render().find(".navbar-item").length).toBe(0);
  });
  it("should render a navbar with pages", () => {
    const pages = ["pageOne", "pageTwo"];
    const wrapper = shallow(<Navbar pages={pages} />);
    expect(wrapper.render().find(".navbar-item").length).toBe(2);
    expect(
      wrapper.find(NavBarLink).everyWhere(node => {
        const text = node.render().text();
        return pages.includes(text);
      })
    ).toBe(true);
  });
  it("should render a navbar with selected pages", () => {
    const pages = ["pageOne", "pageTwo"];
    const wrapper = shallow(<Navbar pages={pages} route={pages[0]} />);
    expect(
      wrapper
        .find(NavBarLink)
        .someWhere(node => node.dive().hasClass("is-active"))
    ).toBe(true);
  });
  it("should handle user clicking on a page", () => {
    const pages = ["pageOne", "pageTwo"];
    const mockedChangeRoute = jest.fn();
    const wrapper = shallow(
      <Navbar pages={pages} changeRoute={mockedChangeRoute} />
    );
    expect(mockedChangeRoute.mock.calls.length).toBe(0);
    wrapper
      .find(NavBarLink)
      .last()
      .dive()
      .simulate("click");
    expect(mockedChangeRoute.mock.calls.length).toBe(1);
  });
  it("should handle user clicking on the hamburger menu", () => {
    const pages = ["pageOne", "pageTwo"];
    const wrapper = shallow(<Navbar pages={pages} />);

    const getBurgerParent = () => wrapper.find(NavBarBurger).parent();
    const renderBurger = () => wrapper.find(NavBarBurger).render();

    expect(renderBurger().hasClass("is-active")).toBe(false);
    getBurgerParent().simulate("click");
    expect(renderBurger().hasClass("is-active")).toBe(true);
    getBurgerParent().simulate("click");
    expect(renderBurger().hasClass("is-active")).toBe(false);
  });
});
