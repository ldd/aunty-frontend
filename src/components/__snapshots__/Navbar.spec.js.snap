// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders snapshot [empty] 1`] = `
<div
  className="hero-head"
>
  <header
    className="navbar"
  >
    <div
      className="container"
    >
      <div
        className="navbar-brand"
        onClick={[Function]}
      >
        <span
          className="navbar-burger burger "
          data-target="navbarMenuHeroC"
        >
          <span />
          <span />
          <span />
        </span>
      </div>
      <div
        className="navbar-menu "
        id="navbarMenuHeroC"
      >
        <div
          className="navbar-end"
        />
      </div>
    </div>
  </header>
</div>
`;

exports[`Navbar renders snapshot [pages] 1`] = `
<div
  className="hero-head"
>
  <header
    className="navbar"
  >
    <div
      className="container"
    >
      <div
        className="navbar-brand"
        onClick={[Function]}
      >
        <span
          className="navbar-burger burger "
          data-target="navbarMenuHeroC"
        >
          <span />
          <span />
          <span />
        </span>
      </div>
      <div
        className="navbar-menu "
        id="navbarMenuHeroC"
      >
        <div
          className="navbar-end"
        >
          <a
            className="navbar-item "
            href="#pageOne"
            onClick={[Function]}
          >
            pageOne
          </a>
          <a
            className="navbar-item "
            href="#pageTwo"
            onClick={[Function]}
          >
            pageTwo
          </a>
        </div>
      </div>
    </div>
  </header>
</div>
`;
